Add thread safety annotations to xds_end2end_test

tamird opened the pull request:
Follow up to #26867.

@yashykt

markdroth commented on test/cpp/end2end/xds_end2end_test.cc:
```diff
@@ -838,6 +839,7 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
                         SubscriptionMap* subscription_map,
                         SentState* sent_state,
                         absl::optional<DiscoveryResponse>* response) {
+      grpc_core::MutexLock lock(&parent_->ads_mu_);
```
This doesn't look right to me.  `ProcessRequest()` is called while already holding the lock, so I think this will deadlock.

I think you just need to tag this method with `ABSL_EXCLUSIVE_LOCKS_REQUIRED`.

tamird replied:
You're absolutely right. Done.

markdroth commented on test/cpp/end2end/xds_end2end_test.cc:
```diff
@@ -941,7 +943,9 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
       gpr_log(GPR_INFO, "ADS[%p]: Received update for type=%s name=%s", this,
               resource_type.c_str(), resource_name.c_str());
       auto& subscription_name_map = (*subscription_map)[resource_type];
+      parent_->ads_mu_.Lock();
```
Same problem here: `ProcessUpdate()` is called when already holding the lock, so this should deadlock.

tamird replied:
Done.

markdroth commented on test/cpp/end2end/xds_end2end_test.cc:
```diff
@@ -997,6 +1001,7 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
         const int version, const SubscriptionNameMap& subscription_name_map,
         const std::set<std::string>& resources_added_to_response,
         SentState* sent_state, DiscoveryResponse* response) {
+      grpc_core::MutexLock lock(&parent_->ads_mu_);
```
Same here.

tamird replied:
Done.

markdroth commented on test/cpp/end2end/xds_end2end_test.cc:
```diff
@@ -644,6 +644,7 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
   }
 
   void NotifyDoneWithAdsCallLocked() {
+    grpc_core::MutexLock lock(&ads_mu_);
```
Looks like this one is also called only when holding the lock, so this is also a deadline.  This method also needs `ABSL_EXCLUSIVE_LOCKS_REQUIRED()`.

tamird replied:
Yup. Done.

markdroth commented:
@veblush Any idea why `ABSL_EXCLUSIVE_LOCKS_REQUIRED()` doesn't seem to be working here?

tamird commented:
I added some `ABSL_ASSERT_EXCLUSIVE_LOCK` and explanatory comments. This smells like a compiler bug to me - I would be happy to report it upstream but don't think it should hold up this PR.

I've also added a few more annotations that I missed the first time around.

tamird commented:
Something is fishy about the fact that the compiler errors mention the template instantiation. There's an attempted reduction in https://gcc.godbolt.org/z/3P17KbvhP which seems to work correctly even when the annotation goes through a pointer, so something else is going on. Again, looks like a compiler bug.

tamird commented on test/cpp/end2end/xds_end2end_test.cc:
```diff
@@ -837,7 +837,17 @@ class AdsServiceImpl : public std::enable_shared_from_this<AdsServiceImpl> {
-                        absl::optional<DiscoveryResponse>* response) {
+                        absl::optional<DiscoveryResponse>* response)
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(parent_->ads_mu_) {
+      // We already told annotalysis that this function requires the lock, but
```
That seems strictly worse. Why throw out the baby with the bathwater? This comment is just a breadcrumb for future readers, and this lambda is a very narrow workaround, where `NO_THREAD_SAFETY_ANALYSIS` is throwing the guardrails out entirely.

 I am currently working on producing a small repro that can be sent upstream as a bug report.

markdroth replied:
I'd really like to hear back from upstream as to what's actually causing this problem before we merge this PR.  I want to make sure we're not somehow misunderstanding something.

tamird replied:
Well, this is looking quite likely to be a compiler bug at this point. https://bugs.llvm.org/show_bug.cgi?id=51368

tamird replied:
I've updated this workaround to be more precise. It will now cause compilation to *fail* if the compiler bug is fixed, which is good because it will prompt a human to come around and remove the workaround.

markdroth replied:
Actually, that may not be ideal, because when the new compiler is released, we'll need the code to work with both the old and new compilers.  Can we do this in a way that will bypass the bug but continue to work when the bug is fixed?

tamird replied:
We certainly can, but do we want to? My worry is that we will simply forget about this code. How about a compromise: I'll add a snippet in this comment that explains what the compiler appeasing alternative would be (probably just a call to AssertHeld) so that the compiler upgrading person can do so.

This won't affect downstream users because this is test code.

markdroth replied:
If the problem is that the compiler isn't able to translate the lock annotations through the template, is there a way that we can make it work via a proxy object *inside* the template?  For example, maybe we could create our a new mutex class inside the template that just delegates all methods to `parent_->ads_mu_`, and then we can have the lock annotations refer to that lock instead of referring directly to `parent_->ads_mu_`?

markdroth replied:
To be clear, I'm really not happy with `AssertHeld()`, because that becomes a run-time error instead of a compile-time error.

tamird replied:
This suggestion won't work because the compiler will then not be able to see that accessing guarded state through `parent_` is permitted.

markdroth commented:
This looks good.  I've triggered the tests, so let's see if everything is happy.

markdroth reviewed (approved):
Just waiting for the last couple of tests to complete.

markdroth commented:
Known issues: #26933 #26595

The "Basic Tests Ruby MacOS" and "Artifact Build MacOS" failures are infrastructure timeouts.
